### PR TITLE
Add simple warning for circular dependencies

### PR DESCRIPTION
**Pull request by adrianheine**
See #1941. This is very primitive and doesn't warn through rollup's logging system.

### CODE REVIEW

File: src/Graph.ts [hunk 1]
```diff
@@ -437,7 +434,9 @@ export default class Graph {
 
 		const visit = (module: Module, seen: { [id: string]: boolean } = {}) => {
 			if (seen[module.id]) {
-				hasCycles = true;
+				if (!allSeen[module.id]) {
+					this.warnCycle(module);
```
Comment (guybedford) [inline src/Graph.ts]: For even better debugging perhaps pass the full cycle stack?

```js
let cycleStack = [];
Object.keys(seen).forEach(name => {
  if (cycleStack) {
    cycleStack.push(name);
  } else if (name === module.id) {
    cycleStack = [];
  }
});
this.warnCycle(module, cycleStack);
```

Then something like `stack: cycleStack.join(' -> ')` in the log.

Comment (guybedford) [reply src/Graph.ts]: Ideally the above should be relative to the base like we do in other places to be readable.

Comment (guybedford) [reply src/Graph.ts]: Sorry, being silly here, this isn't a parent-based structure so doesn't capture the path of course.

**Comment by adrianheine** [comment]
Definitely, this was just a POC.

File: src/Graph.ts (hunk 1 above)
Comment (adrianheine) [reply src/Graph.ts]: I'll try to get the full cycle. I have an idea how to do that :)

Comment (adrianheine) [reply src/Graph.ts]: `relativeId` strips `\0000commonjs-proxy:` prefixes :(

File: src/Graph.ts [hunk 2]
```diff
@@ -428,16 +425,21 @@ export default class Graph {
...
 		const allSeen: { [id: string]: boolean } = {};
 
 		const ordered: Module[] = [];
 
 		const dynamicImports: Module[] = [];
 
+		const cycles: {id: string, path: { [id: string]: string | null }}[] = []
+
 		const visit = (module: Module, seen: { [id: string]: boolean } = {}) => {
+			const modId = relativeId(module.id);
```
Comment (guybedford) [inline src/Graph.ts]: Can we move this into the if statement?

File: src/Graph.ts [hunk 3]
```diff
@@ -428,16 +425,21 @@ export default class Graph {
...
 			if (seen[module.id]) {
-				hasCycles = true;
+				if (!allSeen[module.id]) {
+					cycles.push({id: modId, path: {[modId]: null}});
```
Comment (guybedford) [inline src/Graph.ts]: Spacing - `cycles.push({ id: modId, path: { [modId]: null } });`

File: src/Graph.ts [hunk 4]
```diff
@@ -454,6 +456,15 @@ export default class Graph {
 			module.dependencies.forEach(depModule => {
 				if (!depModule.isExternal) {
 					visit(<Module>depModule, seen);
+					const depId = relativeId(depModule.id);
```
Comment (guybedford) [inline src/Graph.ts]: Same here.

File: src/Graph.ts [hunk 5]
```diff
@@ -490,11 +501,25 @@ export default class Graph {
...
 
-		return { orderedModules: ordered, dynamicImports, hasCycles };
+		cycles.forEach(c => this.warnCycle(c))
+
+		return { orderedModules: ordered, dynamicImports };
 	}
 
-	private warnCycle (_entryModule: Module, _ordered: Module[]) {
-		// TODO: reinstate
+	private warnCycle ({id, path}: {id: string, path: { [id: string]: string | null }}) {
```
Comment (guybedford) [inline src/Graph.ts]: spacing

File: src/Graph.ts [hunk 6]
```diff
@@ -454,6 +456,15 @@ export default class Graph {
 			module.dependencies.forEach(depModule => {
 				if (!depModule.isExternal) {
 					visit(<Module>depModule, seen);
+					const depId = relativeId(depModule.id);
+					cycles.forEach(cycle => {
+						if (cycle.path[depId] === null) {
+							cycle.path[depId] = modId;
```
Comment (guybedford) [inline src/Graph.ts]: `modId` can just be recomputed.

Comment (guybedford) [reply src/Graph.ts]: Alternatively we could do the relative pathing in the warning function itself as a map.

File: src/Graph.ts [hunk 7]
```diff
@@ -454,6 +456,15 @@ export default class Graph {
 			module.dependencies.forEach(depModule => {
 				if (!depModule.isExternal) {
 					visit(<Module>depModule, seen);
+					const depId = relativeId(depModule.id);
+					cycles.forEach(cycle => {
+						if (cycle.path[depId] === null) {
```
Comment (guybedford) [inline src/Graph.ts]: The complexity of this loop does worry me.... cyclical code shouldn't get a performance penalty. Ideally as much of the looping could be moved into the warning function itself, with only markers provided from this main iteration loop (eg parent information etc).

Having these extra loops within the main visit function doesn't seem ideal.

File: src/Graph.ts [hunk 8]
```diff
@@ -454,6 +456,15 @@ export default class Graph {
...
+							cycle.path[depId] = modId;
+							if (cycle.id !== modId) {
+								cycle.path[modId] = null;
+							}
```
Comment (guybedford) [inline src/Graph.ts]: I will have to think about this further as I'm not sure I understand all the edge cases. As I say I'd be much happier merging if we can move as much of this logic into the warning function itself, with simple non-loop-based analysis passed through rather (eg having the `seen` object provide a parent map or similar information that doesn't involve extra looping to create).

Otherwise I'll want to test this much more thoroughly for edge cases and loop complexity.

File: src/Graph.ts (hunk 2 above)
Comment (adrianheine) [reply src/Graph.ts]: It's re-used further down.

File: src/Graph.ts (hunk 4 above)
Comment (adrianheine) [reply src/Graph.ts]: What, moving into the loop?

File: src/Graph.ts (hunk 3 above)
Comment (adrianheine) [reply src/Graph.ts]: Ok.

File: src/Graph.ts (hunk 7 above)
Comment (adrianheine) [reply src/Graph.ts]: This is only looping over the list of cycles, which usually should be 0 or 1.

File: src/Graph.ts (hunk 6 above)
Comment (adrianheine) [reply src/Graph.ts]: Ok. I tried doing it in the warning function, but seemed more complex. I can try again, though.

File: src/Graph.ts (hunk 5 above)
Comment (adrianheine) [reply src/Graph.ts]: Ok.

File: src/Graph.ts (hunk 6 above)
Comment (adrianheine) [reply src/Graph.ts]: Ok, it's actually pretty straightforward.

**Comment by adrianheine** [comment]
Sure, if this is otherwise good to go I can add some tests.

File: src/Graph.ts [hunk 9]
```diff
@@ -454,6 +455,14 @@ export default class Graph {
...
 				if (!depModule.isExternal) {
 					visit(<Module>depModule, seen);
+					cycles.forEach(cycle => {
+						if (cycle.path[depModule.id] === null) {
+							cycle.path[depModule.id] = module.id;
+							if (cycle.id !== module.id) {
+								cycle.path[module.id] = null;
+							}
+						}
+					})
```
Comment (guybedford) [inline src/Graph.ts]: I'd still like to see the complexity reduced here if possible. We are looping each cycle for each dependency for each module and doing a dictionary lookup. A graph like A -> (B,C,D), B -> A, C -> A, D -> ... where the D branch is deep will result in unnecessary looping over the former A,B,C cycles for each module of the D branch, which seems very unnecessary.

Comment (adrianheine) [reply src/Graph.ts]: Just to make sure we have the same understanding of this code: Your example will result in three dictionary lookups per module in the `D` branch, right?

Comment (adrianheine) [reply src/Graph.ts]: I can imagine marking the cycle as closed in L461 (by moving it out of `cycles` into some other array), that would remove unnecessary lookups in your example, but there are probably pathological cases were this wouldn't help at all.

Comment (guybedford) [reply src/Graph.ts]: Yes. This is why I was thinking of parent-based seen as one could do:

```js
if (seen[id]) {
  let cycleStack = [];
  let parentId;
  while (parentId = seen[id]) {
    cycleStack.push(parentId);
    id = parentId;
  }
}
seen[id] = parentId; // with the parent added to the visitor to do this...
```

Each individual `seen` will only have one parent path per module by the nature of `seen`, guaranteeing the above stack which then only iterates the exact path. Although I may be overlooking some of the details.

Comment (adrianheine) [reply src/Graph.ts]: That would make sense, since `seen` is only used as a boolean anyway.

Comment (guybedford) [reply src/Graph.ts]: Note that we may catch cases that aren't actually cycles - eg:

```
A -> B -> C
    -> C
```

C will be in seen, when it is hit again as a parallel dependency not a cyclical one. The way to confirm that it isn't cyclical is that when running up the parent path from the current C (direct dependency of A), we can note that we don't come back around to C. Whether as if we had:

```
C -> A -> B -> C
```

When we come across the second seen "C" and run up the parent stack from the current id, we do find the loop.

Comment (adrianheine) [reply src/Graph.ts]: Are you sure about this? It looks to me like modules are traversed depth-first.

File: src/Graph.ts [hunk 10]
```diff
@@ -453,7 +446,22 @@ export default class Graph {
...
 				if (!depModule.isExternal) {
-					visit(<Module>depModule, seen);
+					if (depModule.id in parents) {
+						if (!allSeen[depModule.id]) {
+							let path = [depModule.id];
+							let id = module.id;
+							while (id !== depModule.id) {
+								path.push(id);
+								id = parents[id];
+								if (!id) break;
```
Comment (adrianheine) [inline src/Graph.ts]: This should never happen imho.

Comment (guybedford) [reply src/Graph.ts]: When the top-level import is hit I think?

Comment (adrianheine) [reply src/Graph.ts]: Yeah, which means there is no cycle, in which case I think this code is not triggered.

Comment (guybedford) [reply src/Graph.ts]: Right of course. How's the output looking?

Comment (adrianheine) [reply src/Graph.ts]: For example like this:

> (!) Circular dependency: node_modules/readable-stream/lib/_stream_readable.js -> node_modules/readable-stream/lib/_stream_duplex.js -> commonjs-proxy:/tmp/rollup-stream-issue-repro/node_modules/readable-stream/lib/_stream_readable.js -> node_modules/readable-stream/lib/_stream_readable.js

Comment (guybedford) [reply src/Graph.ts]: Nice, if we can just get a test with one of these outputs sounds like we're nearly there?

Comment (adrianheine) [reply src/Graph.ts]: I just updated the function tests to expect this warning, so I think that's covered.

Comment (guybedford) [reply src/Graph.ts]: Great I guess we can remove this break then as well?

File: test/function/samples/cycles-pathological-2/_config.js [hunk 11]
```diff
@@ -3,9 +3,21 @@ var assert = require( 'assert' );
...
+		},
+		{
+			code: 'CIRCULAR_DEPENDENCY',
+			importer: 'b.js',
+			message: 'Circular dependency: b.js -> d.js -> c.js -> b.js'
+		},
+		{
+			code: 'CIRCULAR_DEPENDENCY',
+			importer: 'main.js',
+			message: 'Circular dependency: main.js -> b.js -> d.js -> c.js -> main.js'
```
Comment (guybedford) [inline test/function/samples/cycles-pathological-2/_config.js]: Ideally we should avoid warning each module in the cycle I think for readability with these messages?

One option would be to filter by unique cycle here, although that may be quite involved. Another option might be to just permit one cycle warning per module (perhaps keying `cycles` by id and then checking each module against this when populating a cycle path, skipping new cycles that have a match of a module here).

File: src/Graph.ts (hunk 10 above)
Comment (adrianheine) [reply src/Graph.ts]: If my assumption is wrong this will end in an infinite loop, so I would rather like to keep it. Instead of breaking we could just return, though, and maybe emit a warning, since hitting this is a clear sign of a mistake in my reasoning.

File: test/function/samples/cycles-pathological-2/_config.js (hunk 11 above)
Comment (adrianheine) [reply test/function/samples/cycles-pathological-2/_config.js]: I think these are three different cycles. The last two share the `b -> d -> c` part of the path, but otherwise they are different.

File: src/Graph.ts (hunk 10 above)
Comment (guybedford) [reply src/Graph.ts]: I'm still wondering about the case where a imports b and c, and b imports c. As our visit function itself is written as a pre-order visitor, we will call:

* visit(a) for the top-level
* visit(b) for a
* visit(c) for b
* then visit(c) for a again

So I still seem to think that we'd have `c` populated in parents by the time the `visit(c)` for `a` happens. Which then would trigger this case?

File: test/function/samples/cycles-pathological-2/_config.js (hunk 11 above)
Comment (adrianheine) [reply test/function/samples/cycles-pathological-2/_config.js]: I agree, though, that in real-world code cycles seem to overlap a lot or share a root cause and deduplication would be nice.

File: src/Graph.ts (hunk 10 above)
Comment (adrianheine) [reply src/Graph.ts]: No, `visit(c)` for `b` would have finished by then, so that `c` would be in `allSeen`.

File: test/function/samples/cycles-pathological-2/_config.js (hunk 11 above)
Comment (guybedford) [reply test/function/samples/cycles-pathological-2/_config.js]: Ok I see they are fully distinct already so perhaps this won't be necessary. We can always see how it works out in real world cases and if it gets too noisy can then consider such filtering further.

File: src/Graph.ts (hunk 10 above)
Comment (guybedford) [reply src/Graph.ts]: Ahh right. Interesting how allSeen gives us some great higher level properties, which was completely unintended :P

File: src/Graph.ts [hunk 12]
```diff
@@ -453,7 +446,22 @@ export default class Graph {
...
+							}
+							cycles.push({ id: depModule.id, path });
```
Comment (guybedford) [inline src/Graph.ts]: I think we can now move all the logic into the cycle warning function as mentioned before.

Say - `this.warnCycles(id, parents)`, with the logic in there?

Would just be good to keep conceptual boundaries in place.

**Comment by guybedford** [comment]
Just tested this out and it's working perfectly.

In my example building Babel I get 18 cyclical dependency warnings, whereas with the filtering this would be reduced to around 4 or 5 I think.

File: src/Graph.ts (hunk 12 above)
Comment (adrianheine) [reply src/Graph.ts]: `parents` have to be frozen at the time the cycle is detected since there can be a lot of parents per module. I moved the path logic to `warnCycle`, it's not as nice as it was here imho, but probably worth it.

File: src/Graph.ts [hunk 13]
```diff
@@ -453,7 +446,16 @@ export default class Graph {
 
 			module.dependencies.forEach(depModule => {
 				if (!depModule.isExternal) {
-					visit(<Module>depModule, seen);
+					if (depModule.id in parents) {
+						if (!allSeen[depModule.id]) {
+							parents[depModule.id] = module.id;
+							cycles.push({ id: depModule.id, parents: { ...parents } });
```
Comment (guybedford) [inline src/Graph.ts]: Could we not just output the warning directly here instead of collecting the cycles list?

File: src/Graph.ts (hunk 12 above)
Comment (adrianheine) [reply src/Graph.ts]: The last patch doesn't correctly work. Sometimes the cycle doesn't end where it started.

File: src/Graph.ts (hunk 13 above)
Comment (adrianheine) [reply src/Graph.ts]: Sure, I just did that.

File: src/Graph.ts [hunk 14]
```diff
@@ -453,7 +444,15 @@ export default class Graph {
 
 			module.dependencies.forEach(depModule => {
 				if (!depModule.isExternal) {
-					visit(<Module>depModule, seen);
+					if (depModule.id in parents) {
+						if (!allSeen[depModule.id]) {
+							this.warnCycle(depModule.id, { ...parents, [depModule.id]: module.id });
```
Comment (guybedford) [inline src/Graph.ts]: The parents expansion shouldn't be necessary now that this is inline I believe?

Comment (adrianheine) [reply src/Graph.ts]: We may not alter `parents` here because that would break for other cycles involving the same modules. We could do `Object.assign(Object.create(parents}, { [depModule.id]: module.id })`, that's a lot simpler to create but takes more time on lookup.

Comment (guybedford) [reply src/Graph.ts]: But parents isn't changed while `warnCycle` runs synchronously right?

Comment (adrianheine) [reply src/Graph.ts]: No, but we may not set `parents[depModule.id]` either. We could pass `module.id` separately, though.

Comment (guybedford) [reply src/Graph.ts]: Yes that sounds sensible.

Comment (adrianheine) [reply src/Graph.ts]: Did that. Looks better :)

**Comment by guybedford** [comment]
Thanks for your work on this! If there's too much noise, perhaps we could provide an option to turn off these warnings, but lets see how it goes first.

File: test/function/samples/cycles-pathological-2/_config.js (hunk 11 above)
Comment (diegocr) [reply test/function/samples/cycles-pathological-2/_config.js]: Just rechecking https://github.com/rollup/rollup-plugin-commonjs/issues/248 found about this PR which i did overlooked earlier. Great work here, looking forward to full circular references support, though :-)
